File: Csharp/2020/Day02.cs
using System;
using System.IO;
using System.Linq;
using System.Net;
using System.Text.RegularExpressions;


namespace Day02
{
    class Program
    {
        static void Main(string[] args)
        {
            string filePath = "../../data/2020/day02.dat";

            int valid_pass = 0;
            int xor_valid_pass = 0;

            using (StreamReader reader = new(filePath))
            {
                string? line;
                while ((line = reader.ReadLine()) != null)
                {
                    List<string> parts = line.Split(" ").ToList();
                    List<string> low_hi = parts[0].Split("-").ToList();
                    int min_num = int.Parse(low_hi[0]);
                    int max_num = int.Parse(low_hi[1]);
                    char letter = parts[1][0];
                    int num_present = parts[2].Count(c => c == letter);

                    if (num_present >= min_num && num_present <= max_num)
                    {
                        valid_pass++;
                    }

                    if (parts[2][min_num-1] == letter ^ parts[2][max_num-1] == letter)
                    {
                        xor_valid_pass++;
                    }
                }
            }

            Console.WriteLine($"2020 Day 02 Part 1: {valid_pass}");
            Console.WriteLine($"2020 Day 02 Part 2: {xor_valid_pass}");
        }
    }
}

File: java/2020/Day02.java
import module java.base;

public class Day02 {
    private static final Pattern PATTERN = Pattern.compile("(\\d+)-(\\d+) (\\w): (\\w+)");

    public static void main(String[] args) {
        var input = readInput("../../data/2020/day02.dat");
        if (input.isEmpty()) {
            return;
        }

        var data = input.lines().toList();

        System.out.println("Part 1: " + solvePart1(data));
        System.out.println("Part 2: " + solvePart2(data));
    }

    private static String readInput(String filePath) {
        try {
            return Files.readString(Paths.get(filePath)).trim();
        } catch (NoSuchFileException e) {
            System.out.println("Data file missing: " + filePath);
        } catch (IOException ex) {
        }
        return "";
    }

    private record PasswordPolicy(int min, int max, char token, String password) {
        static PasswordPolicy parse(String line) {
            var matcher = PATTERN.matcher(line);
            return matcher.matches()
                ? new PasswordPolicy(
                    Integer.parseInt(matcher.group(1)),
                    Integer.parseInt(matcher.group(2)),
                    matcher.group(3).charAt(0),
                    matcher.group(4))
                : null;
        }

        long countToken() {
            return password.chars().filter(c -> c == token).count();
        }

        boolean isValidPart1() {
            var count = countToken();
            return count >= min && count <= max;
        }

        boolean isValidPart2() {
            return (password.charAt(min - 1) == token) ^ (password.charAt(max - 1) == token);
        }
    }

    private static int solvePart1(List<String> data) {
        return (int) data.stream()
            .map(PasswordPolicy::parse)
            .filter(Objects::nonNull)
            .filter(PasswordPolicy::isValidPart1)
            .count();
    }

    private static int solvePart2(List<String> data) {
        return (int) data.stream()
            .map(PasswordPolicy::parse)
            .filter(Objects::nonNull)
            .filter(PasswordPolicy::isValidPart2)
            .count();
    }
}
